pkg/daemon/ceph/osd/kms/kmip.go

```diff
@@ -0,0 +1,420 @@
+	if err != nil {
+		return "", errors.Wrap(err, "failed to verify kmip register response")
+	}
+
+	var registerRespPayload kmip.RegisterResponsePayload
+	err = decoder.DecodeValue(&registerRespPayload, bi.ResponsePayload.(ttlv.TTLV))
+	if err != nil {
+		return "", errors.Wrap(err, "failed to decode kmip response value")
+	}
+
+	return registerRespPayload.UniqueIdentifier, nil
+}
+
+func (kms *kmipKMS) getKey(uniqueIdentifier string) (string, error) {
+	conn, err := kms.connect()
+	if err != nil {
+		return "", errors.Wrap(err, "failed to connect to kmip kms")
+	}
+	defer conn.Close()
+
+	respMsg, decoder, uniqueBatchItemID, err := kms.send(conn, kmip14.OperationGet, kmip.GetRequestPayload{
+		UniqueIdentifier: uniqueIdentifier,
+	})
+	if err != nil {
+		return "", errors.Wrap(err, "failed to send get request to kmip")
+	}
+	bi, err := kms.verifyResponse(respMsg, kmip14.OperationGet, uniqueBatchItemID)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to verify kmip response")
+	}
+	var getRespPayload kmip.GetResponsePayload
+	err = decoder.DecodeValue(&getRespPayload, bi.ResponsePayload.(ttlv.TTLV))
+	if err != nil {
+		return "", errors.Wrap(err, "failed to decode kmip response value")
+	}
+
+	secretBytes := getRespPayload.SymmetricKey.KeyBlock.KeyValue.KeyMaterial.([]byte)
+	secretBase64 := base64.StdEncoding.EncodeToString(secretBytes)
+
+	return secretBase64, nil
+}
+
+func (kms *kmipKMS) deleteKey(uniqueIdentifier string) error {
+	conn, err := kms.connect()
+	if err != nil {
+		return errors.Wrap(err, "failed to connect to kmip kms")
+	}
+	defer conn.Close()
+
+	respMsg, decoder, uniqueBatchItemID, err := kms.send(conn, kmip14.OperationDestroy, kmip.DestroyRequestPayload{
+		UniqueIdentifier: uniqueIdentifier,
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to send delete request to kmip")
+	}
+	bi, err := kms.verifyResponse(respMsg, kmip14.OperationDestroy, uniqueBatchItemID)
+	if err != nil {
+		return errors.Wrap(err, "failed to verify kmip response")
+	}
+	var destroyRespPayload kmip.DestroyResponsePayload
+	err = decoder.DecodeValue(&destroyRespPayload, bi.ResponsePayload.(ttlv.TTLV))
+	if err != nil {
+		return errors.Wrap(err, "failed to decode kmip response value")
+	}
+
+	return nil
+}
+
+// connect to the kmip endpoint, perform TLS and KMIP handshakes.
+func (kms *kmipKMS) connect() (*tls.Conn, error) {
+	conn, err := tls.Dial("tcp", kms.endpoint, kms.tlsConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial kmip connection endpoint: %w", err)
+	}
+	if kms.readTimeout != 0 {
+		err = conn.SetReadDeadline(time.Now().Add(time.Second * time.Duration(kms.readTimeout)))
+		if err != nil {
+			return nil, fmt.Errorf("failed to set read deadline: %w", err)
+		}
+	}
+	if kms.writeTimeout != 0 {
+		err = conn.SetReadDeadline(time.Now().Add(time.Second * time.Duration(kms.writeTimeout)))
+		if err != nil {
+			return nil, fmt.Errorf("failed to set write deadline: %w", err)
+		}
+	}
+	defer func() {
```

Comment:
Need to move this `defer` up after the `tls.Dial()`? Otherwise, the connectio might not be closed if lines 253 or 259 return errors.

Reply:
I'll do this 👍